hail/python/hail/nd/nd.py

```diff
@@ -2,28 +2,81 @@
 
 import hail as hl
 from hail.expr.functions import _ndarray
+from hail.expr.types import HailType
 from hail.typecheck import *
-from hail.expr.expressions import expr_int64, expr_tuple, expr_any, Int64Expression
+from hail.expr.expressions import expr_int64, expr_tuple, expr_any, Int64Expression, cast_expr
 
 
 def array(input_array):
     return _ndarray(input_array)
 
 
-@typecheck(shape=oneof(expr_int64, tupleof(expr_int64), expr_tuple()), value=expr_any)
-def full(shape, value):
+@typecheck(shape=oneof(expr_int64, tupleof(expr_int64), expr_tuple()), value=expr_any, dtype=nullable(HailType))
+def full(shape, value, dtype=None):
     if isinstance(shape, Int64Expression):
         shape_product = shape
     else:
         shape_product = reduce(lambda a, b: a * b, shape)
-    return array(hl.range(hl.int32(shape_product)).map(lambda x: value)).reshape(shape)
+    return array(hl.range(hl.int32(shape_product)).map(lambda x: cast_expr(value, dtype))).reshape(shape)
```

Comment:
oof! it seems like it might be nice to have something like `hl.array_fill(length, value)` to shortcut this, since it seems pretty widely applicable.

Reply:
Agreed. I'm going to make that an issue for another PR though.